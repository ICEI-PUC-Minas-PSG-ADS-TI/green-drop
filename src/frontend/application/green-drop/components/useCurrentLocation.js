// components/useCurrentLocation.js

import { useState, useEffect, useCallback, useRef } from 'react';
import * as Location from 'expo-location';
import { Alert, Platform } from 'react-native';

const LOCATION_PERMISSION_MESSAGE =
  'Precisamos de acesso à sua localização para fornecer funcionalidades baseadas em GPS.';

export default function useCurrentLocation(options = {}) {
  const subscriberRef = useRef(null);

  // 1) Initial state: location=null, error=null, loading=true
  const [state, setState] = useState({
    location: null,
    error: null,
    loading: true,
  });

  // 2) Pull out any options with defaults
  const {
    enableHighAccuracy = true,
    timeout = 15000,
    maximumAge = 10000,
    watchPosition = false,
    distanceInterval = 10,
    autoRequest = true,
    timeInterval = 5000,
  } = options;

  // 3) Ask the OS for foreground‐location permission
  const requestPermission = useCallback(async () => {
    try {
      const { status } = await Location.requestForegroundPermissionsAsync();
      return status === 'granted';
    } catch (err) {
      // If the native permission API fails, show an alert (tests suppress alerts)
      Alert.alert('Erro de permissão', LOCATION_PERMISSION_MESSAGE);
      return false;
    }
  }, []);

  // 4) The main “get current location” function. It does exactly one setState on success or failure.
  const getCurrentLocation = useCallback(
    async (suppressAlert = false) => {
      try {
        // a) Ask for permission
        const hasPermission = await requestPermission();
        if (!hasPermission) {
          // If permission is denied, set error + loading = false in one shot:
          const errorMessage = 'Permissão negada. ' + LOCATION_PERMISSION_MESSAGE;
          setState({
            location: null,
            error: errorMessage,
            loading: false,
          });
          if (!suppressAlert) {
            Alert.alert('Permissão necessária', LOCATION_PERMISSION_MESSAGE);
          }
          return;
        }

        // b) Permission granted → fetch the GPS position
        const pos = await Location.getCurrentPositionAsync({
          accuracy: Location.Accuracy.High,
          enableHighAccuracy,
          timeout,
          maximumAge,
        });

        // c) Got a position → set it exactly once, with loading=false
        setState({
          location: pos,
          error: null,
          loading: false,
        });
        return pos;
      } catch (err) {
        // d) getCurrentPositionAsync threw (GPS off, timeout, etc.)
        const errorMessage = Platform.select({
          ios: 'Ative os serviços de localização nas configurações.',
          android: 'Habilite a localização precisa nas configurações do dispositivo.',
        });
        setState({
          location: null,
          error: errorMessage,
          loading: false,
        });
        if (!suppressAlert) {
          Alert.alert('Localização indisponível', errorMessage);
        }
        // Re‐throw so tests that expect a rejection can catch it
        throw err;
      }
    },
    [enableHighAccuracy, timeout, maximumAge, requestPermission]
  );

  useEffect(() => {
    // 5) On mount (if autoRequest), immediately call getCurrentLocation(true)
    if (autoRequest) {
      getCurrentLocation(true);
    }

    // 6) If watchPosition = true, subscribe to continuous updates
    if (watchPosition) {
      Location.watchPositionAsync(
        {
          enableHighAccuracy,
          timeInterval,
          distanceInterval,
        },
        newLoc => {
          // Only update “location” (no loading or error)
          setState(prev => ({
            ...prev,
            location: newLoc,
          }));
        }
      )
        .then(sub => {
          subscriberRef.current = sub;
        })
        .catch(err => {
          // If subscribing fails immediately, record error + loading=false
          setState(prev => ({
            ...prev,
            error: err.message,
            loading: false,
          }));
        });
    }

    return () => {
      // 7) On unmount, if we have a subscription, remove it
      if (
        subscriberRef.current &&
        typeof subscriberRef.current.remove === 'function'
      ) {
        subscriberRef.current.remove();
      }
    };
  }, [
    watchPosition,
    distanceInterval,
    autoRequest,
    timeInterval,
    enableHighAccuracy,
    getCurrentLocation,
  ]);

  // 8) Expose exactly what callers/tests expect
  return {
    location: state.location,
    error: state.error,
    loading: state.loading,
    refreshLocation: getCurrentLocation,
    requestPermission,
  };
}
