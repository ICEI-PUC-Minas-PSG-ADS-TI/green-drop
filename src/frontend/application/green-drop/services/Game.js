/*global __DEV__*/
import api from './apiML';

class GameService {
  constructor() {
    this.mock = {
      userStats: { pontos: 150 },
      conquistas: [
        { id: 1, titulo: "Primeiros Passos", descricao: "Faça sua primeira contribuição", completo: true },
        { id: 2, titulo: "Explorador Nato", descricao: "Visite 50 locais diferentes", completo: false },
      ],
      desafios: [
        { id: 1, titulo: "Explorador Iniciante", descricao: "Adicione 5 novos locais", progresso: "3/5", completo: false },
        { id: 2, titulo: "Contribuidor Ativo", descricao: "Faça 10 contribuições este mês", progresso: "7/10", completo: false },
      ],
      historico: [
        {
          id: 1,
          category: 'infra',                    // id da categoria
          status: 'Accepted',                   // situação atual
          problemType: 'Calçadas Danificadas',  // tipo específico
          relevance: 3,                         // nível de relevância (1–5)
          description: 'Calçada quebrada na Rua das Flores, dificultando passagem de pedestres.',
          photo: { uri: 'https://picsum.photos/200/200' }
        },
        {
          id: 2,
          category: 'infra',
          status: 'Pending',
        problemType: 'Buracos na Via',
          relevance: 2,
          description: 'Avenida Principal com vários buracos, risco de acidentes.',
          photo: { uri: 'https://picsum.photos/id/237/200/200' }
        },
        {
          id: 3,
          category: 'servicos',
          status: 'Rejected',
          problemType: 'Quedas de Energia Elétrica',
          relevance: 4,
          description: 'Frequentes quedas de energia no bairro Nova Vista durante a noite.',
          photo: { uri: 'https://picsum.photos/id/297/200/200' }
        },
      ],
    };
  }

  async getUserStats() {
    if (__DEV__) return this.mock.userStats;
    const response = await api.get('/game/userStats');
    return response.data;
  }

  async getConquistas() {
    if (__DEV__) return this.mock.conquistas;
    const response = await api.get('/game/conquistas');
    return response.data;
  }

  async getDesafios() {
    if (__DEV__) return this.mock.desafios;
    const response = await api.get('/game/desafios');
    return response.data;
  }

  async getHistorico() {
    if (__DEV__) return this.mock.historico;
    const response = await api.get('/game/historico');
    return response.data;
  }
}

export const gameService = new GameService();
