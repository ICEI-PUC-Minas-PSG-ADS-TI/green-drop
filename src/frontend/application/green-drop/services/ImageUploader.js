import * as ImagePicker from 'expo-image-picker';
import api from './apiML';

class ImageUploader {
  async pickImage() {
    const { status } = await ImagePicker.requestMediaLibraryPermissionsAsync();
    if (status !== 'granted') throw new Error('Permissão negada');

    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      quality: 0.8,
    });
    if (result.canceled) return null;
    if (!result.assets || result.assets.length === 0) {
      throw new Error('Nenhuma imagem selecionada');
    }
    return result.assets[0];
  }

  async uploadProblemReport(data, image) {
    if (!image?.uri) throw new Error('Imagem inválida');

    const form = new FormData();
    form.append('photo', {
      uri: image.uri,
      name: image.uri.split('/').pop(),
      type: image.type || 'image/jpeg',
    });
    Object.entries(data).forEach(([key, val]) => form.append(key, String(val)));

    const response = await api.post('/reports', form, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Upload falhou: ${response.status} - ${text}`);
    }
    return response.data;
  }
}

export const imageUploader = new ImageUploader();