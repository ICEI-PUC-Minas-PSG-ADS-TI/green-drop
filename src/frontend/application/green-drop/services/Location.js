import api from './apiML';

class LocationService {
  async getNearbyLocations(lat, lng, radius = 5000) {
    try {
      const response = await api.get(`/locations/nearby`, {
        params: {
          latitude: lat,
          longitude: lng,
          radius,
        }
      });
      return response.data.locations;
    } catch (error) {
      // eslint-disable-next-line no-console, no-undef
      console.error('Erro na API de localização:', error);
      throw new Error('Não foi possível buscar locais próximos');
    }
  }
}

export const locationService = new LocationService();