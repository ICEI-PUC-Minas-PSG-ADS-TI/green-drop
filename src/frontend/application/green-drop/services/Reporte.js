/* global __DEV__, FormData */
import api from './apiML';

class ReporteService {
  constructor() {
    this.mockResponse = { reportId: 123, status: 'success', message: 'Relatório enviado (mock)' };
  }

  async uploadProblemReport(data, image) {
    if (__DEV__) return this.mockResponse;
    if (!image?.uri) throw new Error('Imagem inválida');

    const form = new FormData();
    form.append('photo', {
      uri: image.uri,
      name: image.uri.split('/').pop(),
      type: image.type || 'image/jpeg',
    });
    Object.entries(data).forEach(([key, val]) => form.append(key, String(val)));

    const response = await api.post('/reports', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    if (response.status < 200 || response.status >= 300) {
      const text = await response.text();
      throw new Error(`Upload falhou: ${response.status} - ${text}`);
    }
    return response.data;
  }
}

export const reporteService = new ReporteService();