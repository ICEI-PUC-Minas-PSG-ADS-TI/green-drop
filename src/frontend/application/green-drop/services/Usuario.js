import api from "./apiML";

class UsuarioService {
  async login({ email, telefone, senha }) {
    const payload = {
      password: senha,
      ...(email ? { email } : {}),
      ...(telefone ? { phoneNumber: telefone } : {}),
    };
    try {
      const response = await api.post('/users/login', payload);
      if (response.status !== 200) throw new Error('Erro no login');
      return response.data;

    } catch (error) {
      console.error("Erro ao logar:", error);
      return null;

    }
  }

  async cadastro({ nome, email, senha, telefone, role = 'user' }) {
    try {
      const response = await api.post(`/users/${role}`, {
        displayName: nome,
        email,
        password: senha,
        phoneNumber: telefone,
      });
      if (response.status !== 201) throw new Error('Erro no cadastro');
      return response.data;
    } catch (error) {
      console.error("Erro ao cadastrar:", error);
      return null;
    }
  }
}

  export const usuarioService = new UsuarioService();

