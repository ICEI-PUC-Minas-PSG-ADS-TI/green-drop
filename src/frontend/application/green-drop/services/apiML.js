import axios from "axios";
const api = axios.create({
    baseURL: "https://a8fb-186-248-79-98.ngrok-free.app/v1",
});

export default api;