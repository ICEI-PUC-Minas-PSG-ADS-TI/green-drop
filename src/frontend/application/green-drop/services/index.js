export { imageUploader } from './ImageUploader';
export { usuarioService } from './Usuario';
export { locationService } from './Location';