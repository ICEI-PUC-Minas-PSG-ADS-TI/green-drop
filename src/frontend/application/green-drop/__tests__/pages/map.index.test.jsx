// __tests__/pages/map.index.test.jsx

// 1) Mock react-native-maps so <MapView testID="map-view" /> is a simple View with testID
jest.mock('react-native-maps', () => {
  const React = require('react');
  const { View, Text } = require('react-native');
  const MockMapView = ({ children, testID }) => <View testID={testID}>{children}</View>;
  MockMapView.Marker = ({ title }) => <Text>{title}</Text>;
  return {
    __esModule: true,
    default: MockMapView,
  };
});

// 2) Mock expo-location and the custom hook
jest.mock('expo-location');
jest.mock('@/components/useCurrentLocation');

import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react-native';
import MapScreen from '../../pages/map/index.jsx'; // ensure this matches default export
import { LocationService } from '@/services/Location';
import useCurrentLocation from '@/components/useCurrentLocation';

// 3) Stub useNavigation (MapScreen uses navigation.navigate)
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({ navigate: jest.fn() }),
}));

describe('<MapScreen />', () => {
  const fakeCoords = { latitude: -23.5505, longitude: -46.6333 };
  const fakeLocationObj = { coords: fakeCoords };
  const fakeNearbyLocations = [
    { id: '1', name: 'Local Teste 1', latitude: -23.5505, longitude: -46.6333, description: 'Descrição teste', category: 'Cat A' },
    { id: '2', name: 'Local Teste 2', latitude: -23.5600, longitude: -46.6400, description: 'Descrição teste 2', category: 'Cat B' },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
    // Always pretend the hook has resolved
    useCurrentLocation.mockReturnValue({
      location: fakeLocationObj,
      loading: false,
      error: null,
      refreshLocation: jest.fn(),
      requestPermission: jest.fn(),
    });

    // And pretend the service returns two locations
    LocationService.getNearbyLocations.mockResolvedValue(fakeNearbyLocations);
  });

  it('combina com snapshot', () => {
    const tree = render(<MapScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('deve lidar com situação de sucesso', async () => {
    const { getByTestId, getByText } = render(<MapScreen />);
    await waitFor(() => {
      // Now <MapView testID="map-view" /> is rendered by our mock,
      // and the first Marker is a <Text> “Local Teste 1”
      expect(getByTestId('map-view')).toBeTruthy();
      expect(getByText('Local Teste 1')).toBeTruthy();
    });
  });

  it('exibe loading durante busca de locais', async () => {
    // 1) Hook says “done loading” so that MapView / markers can appear…
    useCurrentLocation.mockReturnValue({
      location: fakeLocationObj,
      loading: false,
      error: null,
      refreshLocation: jest.fn(),
      requestPermission: jest.fn(),
    });
    // 2) But service is still pending
    let resolveNearby;
    const pending = new Promise((res) => { resolveNearby = res; });
    LocationService.getNearbyLocations.mockReturnValue(pending);

    const { getByText } = render(<MapScreen />);
    // Since service is pending, isLoading === true, so “Buscando locais…” should show
    expect(getByText('Buscando locais...')).toBeTruthy();

    // Now resolve and wait for markers
    await act(async () => {
      resolveNearby(fakeNearbyLocations);
    });
    await waitFor(() => {
      expect(getByText('Local Teste 1')).toBeTruthy();
    });
  });

  it('mostra erro de localização', () => {
    useCurrentLocation.mockReturnValue({
      location: null,
      loading: false,
      error: 'Verifique as permissões',
      refreshLocation: jest.fn(),
      requestPermission: jest.fn(),
    });

    const { getByText } = render(<MapScreen />);
    expect(getByText(/verifique as permissões/i)).toBeTruthy();
  });

  it('atualiza região ao buscar novos locais', async () => {
    const mockRefresh = jest.fn();
    useCurrentLocation.mockReturnValue({
      location: fakeLocationObj,
      loading: false,
      error: null,
      refreshLocation: mockRefresh,
      requestPermission: jest.fn(),
    });
    LocationService.getNearbyLocations.mockResolvedValue(fakeNearbyLocations);

    const { getByTestId, getByText } = render(<MapScreen />);
    await waitFor(() => {
      expect(getByTestId('map-view')).toBeTruthy();
    });

    fireEvent.press(getByTestId('refresh-button'));
    expect(mockRefresh).toHaveBeenCalledTimes(1);

    await waitFor(() => {
      expect(getByText('Local Teste 1')).toBeTruthy();
    });
  });

  it('mostra detalhes do marcador ao clicar', async () => {
    const { getByText, findByText } = render(<MapScreen />);
    await waitFor(() => {
      expect(getByText('Local Teste 1')).toBeTruthy();
    });
    fireEvent.press(await findByText('Local Teste 1'));
    expect(getByText('Descrição teste')).toBeTruthy();
  });
});
