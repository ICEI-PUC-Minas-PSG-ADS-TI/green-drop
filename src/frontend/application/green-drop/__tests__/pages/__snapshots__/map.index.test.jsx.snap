// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MapScreen /> combina com snapshot 1`] = `null`;
